expan/core/experiment.py

```diff
@@ -1,280 +1,222 @@
-        self.metadata             = metadata.copy()
-        self.report_kpi_names     = report_kpi_names_needed
-        self.derived_kpis         = derived_kpis
-        self.variant_names        = set(self.data.variant)
-        self.control_variant_name = control_variant_name
-        self.reference_kpis       = {}
-
-        # add derived KPIs to the data frame
-        for name, formula in zip(derived_kpi_names, derived_kpi_formulas):
-            self.data.loc[:, name] = eval(re.sub(kpi_name_pattern, r'self.data.\1.astype(float)', formula))
-            self.reference_kpis[name] = re.sub(kpi_name_pattern + '/', '', formula)
-
-    def get_kpi_by_name_and_variant(self, data, name, variant):
-        return data.reset_index().set_index('variant').loc[variant, name]
+    """ Class which adds the analysis functions to experimental data. """
+    def __init__(self, data, metadata):
+        self.data         = data.copy()
+        self.metadata     = metadata.copy()
+        self.worker_table = {
+            'fixed_horizon': statx.make_delta,
+            'group_sequential': es.make_group_sequential,
+            'bayes_factor': es.make_bayes_factor,
+            'bayes_precision': es.make_bayes_precision
+        }
+
 
     def __str__(self):
-        variants = self.variant_names
+        return 'Experiment "{:s}" with {:d} entities.'.format(self.metadata['experiment'], len(self.data))
+
+
+    def analyze_statistical_test(self, test, testmethod, **worker_args):
+        """ Runs delta analysis on one statistical test and returns statistical results.
+        
+        :param test: a statistical test to run
+        :type  test: StatisticalTest
+        :param testmethod: analysis method
+        :type  testmethod: str
+        :param **worker_args: additional arguments for the analysis method
+        :return: statistical result of the test
+        :rtype: StatisticalTestResult
+        """
+        if not isinstance(test, StatisticalTest):
+            raise TypeError("Statistical test should be of type StatisticalTest.")
+
+        if 'entity' not in self.data.columns:
+            raise RuntimeError("There is no 'entity' column in the data.")
+        if self.data.entity.duplicated().any():
+            raise ValueError('Entities in data should be unique.')
+
+        if test.variants.variant_column_name not in self.data.columns:
+            raise RuntimeError("There is no '{}' column in the data.".format(test.variants.variant_column_name))
+        if test.variants.treatment_name not in np.unique(self.data[test.variants.variant_column_name]):
+            raise RuntimeError("There is no treatment with the name '{}' in the data.".format(test.variants.treatment_name))
+        if test.variants.control_name not in np.unique(self.data[test.variants.variant_column_name]):
+            raise RuntimeError("There is no control with the name '{}' in the data.".format(test.variants.control_name))
+
```

Comment:
This really looks very meticulous to me, thanks a lot for this level of attention to the details!